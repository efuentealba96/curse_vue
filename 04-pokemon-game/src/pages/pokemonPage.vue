<template>
  <div class="container-score">
    <h4>Score: {{ score }}</h4>
    <hr>
    <h4>Vidas: {{ lifes }}</h4>
  </div>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Quién es este pokémon?</h1>
    <PokemonPicture v-bind:pokemonID="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions
      :show-list="showList"
      v-bind:pokemons="pokemonArray"
      @selection="checkAnswer($event)"
    />
    <div v-if="showAnswer" class="fade-in">
      <h2>{{ mensaje }}</h2>
      <button v-if="this.lifes !== 0" @click="nextGame">Continuar</button>
      <button v-if="this.lifes === 0" @click="newGame">Nuevo Juego</button>
    </div>
  </div>
</template> 

<script>
import PokemonOptions from "@/components/PokemonOptions.vue";
import PokemonPicture from "@/components/PokemonPicture.vue";

import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  props: {
    lifes: {
      type: Number,
    },
    score: {
      type: Number,
    },
  },
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArray: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      showList: true,
      mensaje: "",
      lifes: 3,
      score: 0,
    };
  },
  methods: {
    async mixPokemonsArray() {
      this.pokemonArray = await getPokemonOptions();

      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArray[rndInt];
    },
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showAnswer = true;
      if (selectedId === this.pokemon.id) {
        this.mensaje = `Correcto el pokemon es ${this.pokemon.name}`;
        this.score += 100;
        this.showList = false;
      } else {
        this.mensaje = `Ops, era ${this.pokemon.name}`;
        this.showList = false;
        this.lifes -= 1;
      }
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArray = [];
      this.showList = true;
      this.lifes = 3;
      this.score = 0;
      this.mixPokemonsArray();
    },
    nextGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArray = [];
      this.showList = true;
      this.mixPokemonsArray();
    },
  },
  mounted() {
    this.mixPokemonsArray();
  },
};
</script>

<style>
.container-score {
  margin-left: 15%;
  width: 10%;
  border: 2px solid green;
}
</style>