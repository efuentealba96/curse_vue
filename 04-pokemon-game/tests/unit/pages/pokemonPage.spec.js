import { shallowMount, mount } from "@vue/test-utils";
import pokemonPage from "@/pages/pokemonPage.vue";
import {pokemons} from "../mocks/pokemon.mock"

describe("PokemonPage Component", () => {
  let wrapper;
  beforeEach(() => {
    
    wrapper = shallowMount(pokemonPage);
  });

  test("debe de llamar el mixPokemon arr al montar", () => {
    const mixPokemonArraySpy = jest.spyOn(
      pokemonPage.methods,
      "mixPokemonsArray"
    );
    shallowMount(pokemonPage);
    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test("debe de hacer match con el snapshot cuando cargan los pokemosn",() =>{
    const wrapper = shallowMount(pokemonPage,{
      data(){
        return {
          pokemonArray: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          showList: true,
          mensaje: "",
          lifes: 3,
          score: 0,
        };
      }
    });

    expect(wrapper.html()).toMatchSnapshot()

  });

  test("debe de mostar los componente de PokemonPicture y PokemonOptions",() =>{
    const wrapper = shallowMount(pokemonPage,{
      data(){
        return {
          pokemonArray: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          showList: true,
          mensaje: "",
          lifes: 3,
          score: 0,
        };
      }
    });
    const picture = wrapper.find("pokemon-picture-stub");
    const options = wrapper.find("pokemon-options-stub")
    expect(picture.exists()).toBeTruthy();
    expect(options.exists()).toBeTruthy();

    expect(picture.attributes("pokemonid")).toBe("1");
    expect(options.attributes("pokemons")).toBeTruthy();
  })




});
