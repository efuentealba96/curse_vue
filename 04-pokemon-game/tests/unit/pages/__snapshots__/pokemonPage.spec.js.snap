// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemosn 1`] = `
<div class="container-score">
  <h4>Score: 0</h4>
  <hr>
  <h4>Vidas: 3</h4>
</div>
<div>
  <h1>¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" showlist="true"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
